distributed/helpers.py

```diff
@@ -31,85 +28,125 @@ def sync_params(model):
                     # use weight of rank 0
                     # important to use copy here otherwise the handle gets detaches from the optimizer
                     param.copy_(tlist[0])
- 
+
 # distributed primitives
-def _reduce(input_, use_fp32=True, group=None):
+# helper routine to compute uneven splitting in balanced way:
+def compute_split_shapes(size, num_chunks):
+    # treat trivial case first
+    if num_chunks == 1:
+        return [size]
+    
+    # first, check if we can split using div-up to balance the load: 
+    chunk_size = (size + num_chunks - 1) // num_chunks
+    last_chunk_size = max(0, size - chunk_size * (num_chunks - 1))
+    if last_chunk_size == 0:
+        # in this case, the last shard would be empty, split with floor instead:
+        chunk_size = size // num_chunks
+        last_chunk_size = size - chunk_size * (num_chunks-1)
+
+    # generate sections list
+    sections = [chunk_size for _ in range(num_chunks - 1)] + [last_chunk_size]
+
+    return sections
+
+def _reduce(input_, comm_name):
     """All-reduce the input tensor across model parallel group."""
-
-    # Bypass the function if we are using only 1 GPU.
-    if dist.get_world_size(group=group) == 1:
+    # Bypass the function if we are using only 1 GPU or if 
+    # communicator is not initialized
+    if comm.get_size(comm_name) == 1:
         return input_
 
     # All-reduce.
-    if use_fp32:
-        dtype = input_.dtype
-        inputf_ = input_.float().contiguous()
-        dist.all_reduce(inputf_, group=group)
-        input_ = inputf_.to(dtype)
-    else:
-        input_ = input_.contiguous()
-        dist.all_reduce(input_, group=group)
+    input_ = input_.contiguous()
+    dist.all_reduce(input_.contiguous(), group=comm.get_group(comm_name))
```

Comment:
The second .contiguous can be dropped here, since input_ is already made config before